src/utils/SubscribableRequestQueue.ts

```diff
@@ -0,0 +1,171 @@
+  private subscribers: Map<number, Map<string, Rejecter>>;
+  /** Map from "inner" request (managed by `queue`) to "outer" promises generated per-subscriber. */
+  private requests: Map<string, RequestSubscription[]>;
+
+  constructor(maxActiveRequests?: number) {
+    this.queue = new RequestQueue(maxActiveRequests);
+    this.subscriberIds = 0;
+    this.subscribers = new Map();
+    this.requests = new Map();
+  }
+
+  /** Resolves all subscriptions to request `key` with `value */
+  private resolveAll<T>(key: string, value: T) {
+    const requests = this.requests.get(key);
+    if (requests) {
+      for (const { resolve, subscriberId } of requests) {
+        resolve(value);
+        this.subscribers.get(subscriberId)?.delete(key);
+      }
+      this.requests.delete(key);
+    }
+  }
+
+  /** Rejects all subscriptions to request `key` with `reason` */
+  private rejectAll(key: string, reason: unknown) {
+    const requests = this.requests.get(key);
+    if (requests) {
+      for (const { reject, subscriberId } of requests) {
+        reject(reason);
+        this.subscribers.get(subscriberId)?.delete(key);
+      }
+      this.requests.delete(key);
+    }
+  }
+
+  /** Adds a new request subscriber. Returns a unique ID to identify this subscriber. */
+  addSubscriber(): number {
+    const subscriberId = this.subscriberIds;
+    this.subscriberIds++;
+    this.subscribers.set(subscriberId, new Map());
+    return subscriberId;
+  }
+
+  /** Queues a new request, or adds a subscription if the request is already queued/running. */
+  addRequestToQueue<T>(
+    key: string,
+    subscriberId: number,
+    requestAction: () => Promise<T>,
+    delayMs?: number
+  ): Promise<T> {
+    // Create single underlying request if it does not yet exist
+    if (!this.queue.hasRequest(key)) {
+      this.queue
+        .addRequest(key, requestAction, delayMs)
+        .then((value) => this.resolveAll(key, value))
+        .catch((reason) => this.rejectAll(key, reason));
+    }
+    if (!this.requests.has(key)) {
+      this.requests.set(key, []);
+    }
+
+    // Validate subscriber
+    if (subscriberId >= this.subscriberIds || subscriberId < 0) {
+      throw new Error(`SubscribableRequestQueue: subscriber id ${subscriberId} has not been registered`);
+    }
+    const subscriber = this.subscribers.get(subscriberId);
+    if (!subscriber) {
+      throw new Error(`SubscribableRequestQueue: subscriber id ${subscriberId} has been removed`);
+    }
+    const existingRequest = subscriber.get(key);
+    if (existingRequest) {
+      this.rejectSubscription(key, existingRequest, "SubscribableRequestQueue: request re-queued while running");
+    }
+
+    // Create promise and add to list of requests
+    return new Promise<T>((resolve, reject) => {
+      this.requests.get(key)?.push({ resolve, reject, subscriberId });
+      this.subscribers.get(subscriberId)?.set(key, reject);
+    });
```

Comment:
Does rejecting and then re-adding the subscription have any consequences for the caller? Would it be better to just return the same promise as before?

Reply:
Probably worth documenting this behavior in the method comment as well